packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,72 +166,228 @@ function makeSwingStore(dirPath, forceReset = false) {
+      const innerReader = new Readlines(fd);
+      function* reader() {
+        try {
+          while (true) {
+            const statusInner = streamStatus.get(streamName);
+            assert(
+              statusInner === readStatus,
+              X`can't read stream ${q(streamName)}, it's been closed`,
+            );
+            const line = /** @type {string|false} */ (innerReader.next());
+            if (line) {
+              itemCount -= 1;
+              assert(itemCount >= 0);
+              const result = line.toString();
+              if (skipCount > 0) {
+                skipCount -= 1;
+              } else {
+                yield result;
+              }
+            } else {
+              break;
+            }
           }
+        } catch (e) {
+          console.log(e);
+        } finally {
+          const statusEnd = streamStatus.get(streamName);
+          assert(
+            statusEnd === readStatus,
+            X`can't read stream ${q(streamName)}, it's been closed`,
+          );
+          closeStream(streamName);
+          assert(itemCount === 0, X`leftover item count ${q(itemCount)}`);
         }
-        fs.closeSync(fd);
-      }
-    } catch (e) {
-      if (e.code !== 'ENOENT') {
-        throw e;
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
+   *
+   * @param {string} streamName  The stream to be written
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @returns {StreamWriter} a writer for the named stream
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      !status,
+      X`can't write stream ${q(streamName)} because it's already in use`,
     );
-    assert(streamName.match(/^[-\w]+$/));
-    let fd = streamFds.get(streamName);
-    if (!fd) {
-      fd = fs.openSync(`${dirPath}/${streamName}`, 'a');
-      streamFds.set(streamName, fd);
+    streamStatus.set(streamName, 'write');
+
+    // XXX fdTemp is a workaround for a flaw in TypeScript's type inference
+    // It should be fd, which it should be changed to when and if they fix tsc.
+    let fdTemp = streamFds.get(streamName);
+    if (!fdTemp) {
+      const filePath = `${dirPath}/${streamName}`;
+      const mode = fs.existsSync(filePath) ? 'r+' : 'w';
+      fdTemp = fs.openSync(filePath, mode);
+      streamFds.set(streamName, fdTemp);
     }
+    const fd = fdTemp;
     activeStreamFds.add(fd);
-    fs.writeSync(fd, item);
-    fs.writeSync(fd, '\n');
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
```

Comment:
What do you think about making this position mandatory? I think that might be slightly less error-prone, and I expect the caller (VatWarehouse/transcript-append) to always have a `position` to offer, since it will always read it out of the kvStore just before calling write().

Reply:
Yes.  I think that might have just been an oversight in the typedef.

Reply:
No, it wasn't.  But the use pattern works better with always providing the position, because you can just initialize the position to `STREAM_START` and then you don't have to have the first write call be treated special in the code.  I think this method was designed before I provided `STREAM_START` as part of the API.